### PR TITLE
Update NNPI Backend to v0.5.0.6

### PR DESCRIPTION
**Pull request by omromano**
Summary: Updating NNPI Backend to version 0.5.0.6.

### CODE REVIEW

**Comment by jsubag** [comment]
@arunm-git @jfix71 Please take a look.

File: lib/Backends/NNPI/tests/TestBlackList.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  auto useInfAPI = getenv("USE_INF_API");
+  if (useInfAPI && !strcmp(useInfAPI, "1")) {
+    return NNPI_EXECUTION_ENGINE_HW;
+  }
+  return NNPI_EXECUTION_ENGINE_SW;
+}
+
+typedef std::pair<std::string, uint32_t> TestSetup;
+
+void prepareBlacklist(const std::vector<TestSetup> testBlacklistedSetups,
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/TestBlackList.h]: nit: make this a reference

File: lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp [hunk 2]
```diff
@@ -12,45 +12,38 @@
...
-          "convOps/0",
-          "localResponseNormalizationTest/0",
-          "tinyResnet/0",
-      });
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp [hunk 3]
```diff
@@ -12,45 +12,38 @@
...
-      // be subject to the static initialization order fiasco.
-      backendTestBlacklist.insert({
-          "convOps/0",
-          "localResponseNormalizationTest/0",
-          "tinyResnet/0",
-      });
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp [hunk 4]
```diff
@@ -12,29 +12,22 @@
...
-      // backendTestBlacklist because they are ordered within the same
-      // translation unit (this source file).  Otherwise this technique would
-      // be subject to the static initialization order fiasco.
-      backendTestBlacklist.erase({"TransferStaticPlaceholderTest/0"});
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp [hunk 5]
```diff
@@ -12,29 +12,22 @@
...
-    if (useInfAPI && !strcmp(useInfAPI, "1")) {
-      // N.B.: This insertion is defined to come after the initialization of
-      // backendTestBlacklist because they are ordered within the same
-      // translation unit (this source file).  Otherwise this technique would
-      // be subject to the static initialization order fiasco.
-      backendTestBlacklist.erase({"TransferStaticPlaceholderTest/0"});
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIMLTest.cpp [hunk 6]
```diff
@@ -12,26 +12,38 @@
...
-    "convNetForImageRecognition/0",
-    "testFindPixelRegression/0",
-    "matrixRotationRecognition/0",
-    "learnSparseLengthsWeightedSumEmbeddings/0",
-    "learnSparseLengthsWeightedSumWeights/0"};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIMLTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIMLTest.cpp [hunk 7]
```diff
@@ -12,26 +12,38 @@
...
-    "learnSparseLengthsSumEmbeddings/0",
-    "nonLinearClassifier/0",
-    "convNetForImageRecognition/0",
-    "testFindPixelRegression/0",
-    "matrixRotationRecognition/0",
-    "learnSparseLengthsWeightedSumEmbeddings/0",
-    "learnSparseLengthsWeightedSumWeights/0"};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIMLTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp [hunk 8]
```diff
@@ -12,11 +12,19 @@
...
 using namespace glow;
 
-std::set<std::string> glow::backendTestBlacklist = {
-    "fc/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp [hunk 9]
```diff
@@ -12,11 +12,19 @@
...
 #include "tests/unittests/BackendTestUtils.h"
 
 using namespace glow;
 
-std::set<std::string> glow::backendTestBlacklist = {
-    "fc/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIOperatorTest.cpp [hunk 10]
```diff
@@ -13,181 +13,217 @@
...
-          "SoftMax/0",
-          "TopK/0",
-          "TopK1/0",
-      });
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIOperatorTest.cpp [hunk 11]
```diff
@@ -13,181 +13,217 @@
...
-          "less_int32Cases/0",
-          "less_int64Cases/0",
-          "SoftMax/0",
-          "TopK/0",
-          "TopK1/0",
-      });
-    }
-  }
-} emuTests;
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp [hunk 12]
```diff
@@ -13,119 +13,201 @@
...
-        "BatchMatMulTest_Int8/46", "BatchMatMulTest_Int8/51",
-        "BatchMatMulTest_Int8/54", "BatchMatMulTest_Int8/59",
-        "BatchMatMulTest_Int8/63", "BatchMatMulTest_Int8/67",
-        "BatchMatMulTest_Int8/71", "BatchMatMulTest_Int8/75",
-        "BatchMatMulTest_Int8/83", "BatchMatMulTest_Int8/87",
-        "BatchMatMulTest_Int8/90", "BatchMatMulTest_Int8/91",
-        "BatchMatMulTest_Int8/95", "FCTest_Int8/139",
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp [hunk 13]
```diff
@@ -13,119 +13,201 @@
...
-    backendTestBlacklist = {
-        "BatchMatMulTest_Int8/39", "BatchMatMulTest_Int8/43",
-        "BatchMatMulTest_Int8/46", "BatchMatMulTest_Int8/51",
-        "BatchMatMulTest_Int8/54", "BatchMatMulTest_Int8/59",
-        "BatchMatMulTest_Int8/63", "BatchMatMulTest_Int8/67",
-        "BatchMatMulTest_Int8/71", "BatchMatMulTest_Int8/75",
-        "BatchMatMulTest_Int8/83", "BatchMatMulTest_Int8/87",
-        "BatchMatMulTest_Int8/90", "BatchMatMulTest_Int8/91",
-        "BatchMatMulTest_Int8/95", "FCTest_Int8/139",
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp [hunk 14]
```diff
@@ -12,15 +12,28 @@
...
-    "multipleBackingTensors/0",  "internalGap/0",
-    "onlyTraceEvents/0",         "twoCompiles/0",
-    "manualAndAutomatic/0",      "incompleteCoverage/0",
-    "deviceManagerEvents/0",     "automaticInstrumentation/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp [hunk 15]
```diff
@@ -12,15 +12,28 @@
...
-std::set<std::string> glow::backendTestBlacklist = {
-    "multipleRunsAreDistinct/0", "manualEvents/0",
-    "multipleBackingTensors/0",  "internalGap/0",
-    "onlyTraceEvents/0",         "twoCompiles/0",
-    "manualAndAutomatic/0",      "incompleteCoverage/0",
-    "deviceManagerEvents/0",     "automaticInstrumentation/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp [hunk 16]
```diff
@@ -12,26 +12,34 @@
...
-    "gradientCheckFCConcatTanh/0",
-    "gradientCheckFC/0",
-    "gradientCheckTranspose/0",
-    "gradientCheckCrossEntropyLoss/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp [hunk 17]
```diff
@@ -12,26 +12,34 @@
...
-    "gradientCheckArithmeticDiv/0",
-    "gradientCheckArithmetic/0",
-    "gradientCheckFCConcatTanh/0",
-    "gradientCheckFC/0",
-    "gradientCheckTranspose/0",
-    "gradientCheckCrossEntropyLoss/0",
-};
+std::set<std::string> glow::backendTestBlacklist = {};
+
+struct Blacklist_Initializer {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp (hunk 14 above)
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp (hunk 15 above)
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp]: nit: `BlacklistInititalizer`

File: lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp (hunk 9 above)
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp]: nit: `BlacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp (hunk 8 above)
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp]: nit: `const std::vector<TestSetup>`

File: lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp [hunk 18]
```diff
@@ -12,45 +12,38 @@
...
+        {"quantizedConvTest/0", TestBlackList::AnyDeviceAnyEngine},
+        {"softmaxGradTest/0", TestBlackList::AnyDeviceAnyEngine},
+        {"convOps/0", TestBlackList::AnyDeviceEngineHW},
+        {"localResponseNormalizationTest/0", TestBlackList::AnyDeviceEngineHW},
+        {"tinyResnet/0", TestBlackList::AnyDeviceEngineHW},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIBackendCorrectnessTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp [hunk 19]
```diff
@@ -12,29 +12,22 @@
...
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
+        {"MultiFunction/0", TestBlackList::AnyDeviceAnyEngine},
+        {"DeviceResidentTensors/0", TestBlackList::AnyDeviceAnyEngine},
+        {"TransferStaticPlaceholderTest/0", TestBlackList::AnyDeviceEngineSW},
+        {"CanHandleDeviceResidentTensors/0", TestBlackList::AnyDeviceAnyEngine},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIDeviceManagerTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp [hunk 20]
```diff
@@ -12,26 +12,34 @@
...
+        {"gradientCheckArithmetic/0", TestBlackList::AnyDeviceAnyEngine},
+        {"gradientCheckFCConcatTanh/0", TestBlackList::AnyDeviceAnyEngine},
+        {"gradientCheckFC/0", TestBlackList::AnyDeviceAnyEngine},
+        {"gradientCheckTranspose/0", TestBlackList::AnyDeviceAnyEngine},
+        {"gradientCheckCrossEntropyLoss/0", TestBlackList::AnyDeviceAnyEngine},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIGradCheckTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIMLTest.cpp [hunk 21]
```diff
@@ -12,26 +12,38 @@
...
+        {"matrixRotationRecognition/0", TestBlackList::AnyDeviceAnyEngine},
+        {"learnSparseLengthsWeightedSumEmbeddings/0",
+         TestBlackList::AnyDeviceAnyEngine},
+        {"learnSparseLengthsWeightedSumWeights/0",
+         TestBlackList::AnyDeviceAnyEngine},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIMLTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp [hunk 22]
```diff
@@ -12,11 +12,19 @@
...
+  Blacklist_Initializer() {
+    std::vector<std::pair<std::string, uint32_t>> testBlacklistedSetups = {
+        {"fc/0", TestBlackList::AnyDeviceAnyEngine},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorGradTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIOperatorTest.cpp [hunk 23]
```diff
@@ -13,181 +13,217 @@
...
+        {"nms_v4_center_point_box_with_gather_float/0",
+         TestBlackList::AnyDeviceEngineHW},
+        {"SoftMax/0", TestBlackList::AnyDeviceEngineHW},
+        {"TopK/0", TestBlackList::AnyDeviceEngineHW},
+        {"TopK1/0", TestBlackList::AnyDeviceEngineHW},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIOperatorTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp [hunk 24]
```diff
@@ -13,119 +13,201 @@
...
+    if (useNnpiThreshold && strcmp(useNnpiThreshold, "1") == 0) {
+      testBlacklistedSetups.clear();
     }
-  }
-} blacklistInitializer;
+
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPIParameterSweepTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp [hunk 25]
```diff
@@ -12,15 +12,28 @@
...
+        {"twoCompiles/0", TestBlackList::AnyDeviceAnyEngine},
+        {"manualAndAutomatic/0", TestBlackList::AnyDeviceAnyEngine},
+        {"incompleteCoverage/0", TestBlackList::AnyDeviceAnyEngine},
+        {"deviceManagerEvents/0", TestBlackList::AnyDeviceAnyEngine},
+        {"automaticInstrumentation/0", TestBlackList::AnyDeviceAnyEngine},
+    };
+    TestBlackList::prepareBlacklist(testBlacklistedSetups,
+                                    backendTestBlacklist);
+  };
+} blacklist_Initializer;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/NNPITraceEventsTest.cpp]: nit: `blacklistInitializer`

File: lib/Backends/NNPI/tests/TestBlackList.h [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cstring>
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+namespace TestBlackList {
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/TestBlackList.h]: nit: For consistency with everywhere else, I'd prefer `TestBlacklist` (i.e. don't capitalize `list`). Could rename the file too.

File: lib/Backends/NNPI/tests/TestBlackList.h [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+  NNPI_EXECUTION_ENGINE_HW = 0b10 << 16,
+  NNPI_EXECUTION_ENGINE_ANY = 0b11 << 16,
+};
+
+const uint32_t AnyDeviceAnyEngine =
+    NNPI_DEVICE_VERSION_ANY | NNPI_EXECUTION_ENGINE_ANY;
+const uint32_t AnyDeviceEngineHW =
+    AnyDeviceAnyEngine ^ NNPI_EXECUTION_ENGINE_SW;
+const uint32_t AnyDeviceEngineSW =
+    AnyDeviceAnyEngine ^ NNPI_EXECUTION_ENGINE_HW;
```
Comment (jfix71) [inline lib/Backends/NNPI/tests/TestBlackList.h]: Naming here is a bit confusing -- "AnyDeviceAnyEngine" vs. "AnyDeviceEngineHW". I'd suggest "AnyDeviceHWEngine"

**Pull request closed without merging**